### DISCUSSION TITLE
KANBAN-Style UI's In SLINT?

### DISCUSSION THREAD

**Discussion by PackMeister** [slint-ui/slint#5779] [Q&A]
Has anyone experience using SLINT to implement a **_KANBAN-style_** interface?
Specifically, we'd like the UI to have a series of columns where each column represented a stage in some process pipeline, e.g. sales, where each column would contain the people or entity in that particular stage, e.g. prospect.
We'd need the UI to permit users to drag-and-drop people or entities **between** columns and **within** a given column.
Is this something SLINT can support on a dedicated, offline-first desktop application?

**Comment by Vadoola** [comment]
I'm sure you could. There is a discussion [here](https://github.com/slint-ui/slint/discussions/5248) about trying to create a node graph style editor. You could probably do something similar with cards and based on some sort of tag in your model manually snap them to columns? They could be dragged around like in that node graph example but you would need to add some logic for snapping it to a column.

If you don't care about dragging the kanban cards around you could probably do it pretty easily with a model per column and multiple vertical layouts. Just updating the models as needed. You might even be able to do something use touch handlers and popup windows to drag around and update the models in vertical layouts but I'm not sure that would work.

**Reply by PackMeister** [reply]
Is there any documentation or especially diagrams that would help me understand the hierarchy SLINT uses to build UIs?

Is the foundational 'object' a window, that 'contains' one or more 'layouts'?

Are horizontal and vertical layouts nestable?

**Reply by Vadoola** [reply]
https://releases.slint.dev/

> Is the foundational 'object' a window, that 'contains' one or more 'layouts'?

Yes

> Are horizontal and vertical layouts nestable?

Yes

**Reply by PackMeister** [reply]
Thanks!   

Even without explicit diagrams, I've found going through some of the exercises has begun to give me a better and clearer idea of how SLINT's layout architecture works.

**Comment by PackMeister** [comment]
The Shared List example from [SortableJS](https://sortablejs.github.io/Sortable/#comparisons) illustrates what I'd like to be able to create in SLINT.

**Reply by Vadoola** [reply]
I don't have the time to try and work out a fully functional example for you, but here is something I quickly hacked together. The basics are there, but the snapping doesn't fully work, and things don't change places, just the square being moved. Hopefully it should at least point you in the right direction, and show you what is possible. I'm also not saying this is the ideal way to do it, like I said just something I quicly threw together. [Slintpad link](https://slintpad.com/?snippet=import+%7B+AboutSlint%2C+Button%2C+VerticalBox%2C+ScrollView+%7D+from+%22std-widgets.slint%22%3B%0A%0A%0Acomponent+card+inherits+Rectangle+%7B%0A++++function+get_x%28idx%3A+int%29+-%3E+length+%7B%0A++++++++%28mod%28idx%2C+4%29+*+105px%29%0A++++%7D%0A%0A++++function+get_y%28idx%3A+int%29+-%3E+length+%7B%0A++++++++%28floor%28idx+%2F+4%29+*+105px%29%0A++++%7D%0A++++%0A++++in+property+%3Cstring%3E+label%3B%0A++++in-out+property+%3Cint%3E+index%3B%0A++++height%3A+100px%3B%0A++++width%3A+100px%3B%0A%0A++++x%3A+get_x%28self.index%29%3B%0A++++y%3A+get_y%28self.index%29%3B%0A%0A++++border-width%3A+2px%3B%0A++++border-color%3A+black%3B%0A++++background%3A+ta.pressed+%3F+lightblue+%3A+lightgray%3B%0A++++Text+%7B%0A++++++++color%3A+black%3B%0A++++++++font-size%3A+20px%3B%0A++++++++horizontal-alignment%3A+center%3B%0A++++++++vertical-alignment%3A+center%3B%0A++++++++text%3A+label%3B%0A++++++++%2F%2Ftext%3A++%22%5C%7Broot.index%7D+%2F+%5C%7Broot.x%2F1px%7D%22%3B%0A++++%7D%0A++++ta+%3A%3D+TouchArea+%7B%0A++++++++height%3A+100%25%3B%0A++++++++width%3A+100%25%3B%0A++++++++moved+%3D%3E+%7B%0A++++++++++++if+self.pressed+%7B%0A++++++++++++++++root.z+%3D+100.0%3B%0A++++++++++++++++root.x+%2B%3D+self.mouse-x+-+self.pressed-x%3B%0A++++++++++++++++root.y+%2B%3D+self.mouse-y+-+self.pressed-y%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++++++pointer-event%28pe%29+%3D%3E+%7B%0A++++++++++++if+pe.kind+%3D%3D+PointerEventKind.up+%7B%0A++++++++++++++++if+root.x+%3E+%28mod%28root.index%2C+4%29+*+50px%29+%7B%0A++++++++++++++++++++root.index+%3D+root.index+%2B+1%3B%0A++++++++++++++++++++root.x+%3D+get_x%28root.index%29%3B%0A++++++++++++++++%7D+else+if+root.x+%3C+%28mod%28root.index%2C+4%29+*+50px%29+%7B%0A++++++++++++++++++++root.index+%3D+root.index+-+1%3B%0A++++++++++++++++++++root.x+%3D+get_x%28root.index%29%3B%0A++++++++++++++++%7D%0A++++++++++++++++%2F*if+root.y+%3E+%28mod%28root.index%2C+4%29+*+50px%29+%7B%0A++++++++++++++++++++root.index+%3D+root.index+%2B+4%3B%0A++++++++++++++++++++root.y+%3D+get_y%28root.index%29%3B%0A++++++++++++++++%7D+else+if+root.y+%3C+%28mod%28root.index%2C+4%29+*+50px%29+%7B%0A++++++++++++++++++++root.index+%3D+root.index+-+4%3B%0A++++++++++++++++++++root.y+%3D+get_y%28root.index%29%3B%0A++++++++++++++++%7D*%2F%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D%0Aexport+component+Demo+%7B%0A++++function+get_x%28idx%3A+int%29+-%3E+length+%7B%0A++++++++%28mod%28idx%2C+4%29+*+105px%29%0A++++%7D%0A%0A++++function+get_y%28idx%3A+int%29+-%3E+length+%7B%0A++++++++%28floor%28idx+%2F+4%29+*+105px%29%0A++++%7D%0A++++%0A++++%0A++++ScrollView+%7B%0A++++++++for+lbl%5Bidx%5D+in+%5B%0A++++++++++++1%2C2%2C3%2C4%2C5%2C6%2C7%2C8%2C9%2C10%2C11%2C12%0A++++++++%5D+%3A+Rectangle+%7B%0A++++++++++++height%3A+100px%3B%0A++++++++++++width%3A+100px%3B%0A++++++++++++Rectangle+%7B%0A++++++++++++++++height%3A+100px%3B%0A++++++++++++++++width%3A+100px%3B%0A++++++++++++++++x%3A+get_x%28idx%29%3B%0A++++++++++++++++y%3A+get_y%28idx%29%3B%0A++++++++++++++++border-width%3A+2px%3B%0A++++++++++++++++border-color%3A+red%3B%0A++++++++++++%7D%0A++++++++++++card+%7B%0A++++++++++++++++index%3A+idx%3B%0A++++++++++++++++%2F%2Fx%3A+get_x%28self.index%29%3B%0A++++++++++++++++%2F%2Fy%3A+get_y%28self.index%29%3B%0A++++++++++++++++label%3A+lbl%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%7D%0A&style=native)
